File: platformio/servo/device_kit/src/As5600.cpp
#include <device_kit/As5600.hpp>

#include <stdexcept>
#include <utility>
#include <cmath>


namespace device_kit
{
    As5600::As5600(bus_io::I2cSlave&& slave)
    :   slave_ {std::move(slave)}
    {
    }


    void As5600::configure(As5600Settings const&)
    {
        // slave_.writeWord(reg_config, r);
    }


    As5600Angle As5600::readAngle()
    {
        std::uint8_t angle_registers[4];
        slave_.read(regRawAngle_, angle_registers);

        As5600Angle angle;
        angle.rawAngle = static_cast<std::uint16_t>(angle_registers[0] << 8 | angle_registers[1]);
        angle.angle = static_cast<std::uint16_t>(angle_registers[2] << 8 | angle_registers[3]);
        angle.angleRadians = 2.f * M_PIf * angle.angle / (1U << bits_);

        return angle;
    }


    As5600Status As5600::readStatus()
    {
        std::uint8_t status_registers[4];
        slave_.read(regStatus_, status_registers);

        return As5600Status {
            .magnetDetected = (status_registers[0] & 0b00100000) != 0,
            .magnetTooWeak = (status_registers[0] & 0b00010000) != 0,
            .magnetTooStrong = (status_registers[0] & 0b00001000) != 0,
            .automaticGainControl = status_registers[1],
            .magnitude = static_cast<std::uint16_t>(status_registers[2] << 8 | status_registers[3])
        };
    }
}

File: src/drivers/i2c/AS5600.js

const STATUS    = 0x0b;
const RAW_ANGLE = 0x0c;
const ANGLE     = 0x0e;


export function AS5600(device) {
    return {
        typical: [0x36],
        reader: device.blockReader,
        angle: function(reg=0x0c) {
            var buf = new Uint8Array(2);
            this.reader(RAW_ANGLE, buf.buffer);
            return (buf[0]<<8)|buf[1];
        },
        scaledAngle: function() {
            return angle(0x0e);
        },
        status: function() {
            var status_registers = new Uint8Array(4);
            this.reader(STATUS, status_registers.buffer);
            return {
                magnetDetected: (status_registers[0] & 0b00100000) != 0,
                magnetTooWeak: (status_registers[0] & 0b00010000) != 0,
                magnetTooStrong: (status_registers[0] & 0b00001000) != 0,
                automaticGainControl: status_registers[1],
                magnitude: status_registers[2] << 8 | status_registers[3]
            }
        }
    };
}

